File: python-backend/old_files/matcher.py
"""事件和市场匹配引擎

匹配逻辑优化:
- 事件匹配: 使用队伍缩写 + 开赛日期匹配两个平台的事件
- 市场匹配: 2:1 匹配（两个 Kalshi 市场对应一个 Poly 市场）

关键点:
- Kalshi: 一个事件有 2 个市场（每个队伍一个）
- Polymarket: 一个事件只有 1 个市场（包含两个队伍的价格）
- 匹配时根据队伍名称确定 Poly 的价格视角
"""
import logging
from typing import List, Dict, Optional, Tuple
from datetime import datetime
from models import (
    KalshiEvent, KalshiMarket, 
    PolymarketEvent, PolymarketMarket,
    MatchedEvent, MatchedMarket
)

logger = logging.getLogger(__name__)


class EventMatcher:
    """事件匹配器 - 两阶段匹配"""
    
    def __init__(self, time_tolerance_hours: int = 24):
        self.time_tolerance_hours = time_tolerance_hours
    
    def match_events_and_markets(
        self,
        kalshi_events: List[KalshiEvent],
        kalshi_markets: List[KalshiMarket],
        polymarket_events: List[PolymarketEvent],
        polymarket_markets: List[PolymarketMarket]
    ) -> Tuple[List[MatchedEvent], List[MatchedMarket]]:
        """执行两阶段匹配"""
        logger.info("=" * 60)
        logger.info("🔍 开始两阶段匹配 (不拆分版)")
        logger.info(f"   Kalshi: {len(kalshi_events)} 个事件, {len(kalshi_markets)} 个市场")
        logger.info(f"   Polymarket: {len(polymarket_events)} 个事件, {len(polymarket_markets)} 个市场")
        logger.info("=" * 60)
        
        # 第一阶段: 事件匹配
        matched_events = self._match_events(kalshi_events, polymarket_events)
        logger.info(f"📊 第一阶段完成: 匹配到 {len(matched_events)} 个事件")
        
        # 第二阶段: 市场匹配 (2:1)
        matched_markets = self._match_markets(matched_events)
        logger.info(f"📊 第二阶段完成: 匹配到 {len(matched_markets)} 个市场对")
        
        return matched_events, matched_markets
    
    def _match_events(
        self,
        kalshi_events: List[KalshiEvent],
        polymarket_events: List[PolymarketEvent]
    ) -> List[MatchedEvent]:
        """第一阶段: 事件匹配"""
        logger.info("-" * 40)
        logger.info("🎯 第一阶段: 事件匹配")
        logger.info("-" * 40)
        
        matched_events = []
        used_poly_ids = set()
        
        # 构建 Polymarket 事件索引: event_name -> [events]
        poly_index: Dict[str, List[PolymarketEvent]] = {}
        for event in polymarket_events:
            key = event.name.upper()
            if key not in poly_index:
                poly_index[key] = []
            poly_index[key].append(event)
        
        # 为每个 Kalshi 事件寻找匹配
        for k_event in kalshi_events:
            k_name = k_event.name.upper()
            k_date = k_event.start_time.date() if k_event.start_time else None
            
            # 也检查反向名称 (如 MEM-LAL vs LAL-MEM)
            parts = k_name.split('-')
            k_name_reversed = f"{parts[1]}-{parts[0]}" if len(parts) == 2 else None
            
            best_match: Optional[PolymarketEvent] = None
            best_confidence = 0.0
            
            # 查找精确匹配或反向匹配
            for name_to_check, is_reversed in [(k_name, False), (k_name_reversed, True)]:
                if not name_to_check:
                    continue
                
                candidates = poly_index.get(name_to_check, [])
                
                for p_event in candidates:
                    if p_event.event_id in used_poly_ids:
                        continue
                    
                    p_date = p_event.start_time.date() if p_event.start_time else None
                    
                    # 验证日期
                    if k_date and p_date:
                        if k_date != p_date:
                            logger.debug(f"   ❌ 日期不匹配: {k_event.name} ({k_date}) vs {p_event.name} ({p_date})")
                            continue
                        confidence = 1.0 if not is_reversed else 0.95
                    else:
                        logger.warning(f"   ⚠️ 缺少日期: {k_event.name} ({k_date}) vs {p_event.name} ({p_date})")
                        confidence = 0.7 if not is_reversed else 0.65
                    
                    if confidence > best_confidence:
                        best_confidence = confidence
                        best_match = p_event
            
            if best_match and best_confidence >= 0.7:
                matched = MatchedEvent(
                    event_name=k_event.name,
                    kalshi_event=k_event,
                    polymarket_event=best_match,
                    confidence=best_confidence
                )
                matched_events.append(matched)
                used_poly_ids.add(best_match.event_id)
                
                logger.info(f"   ✅ 匹配: {k_event.name} <-> {best_match.name} (置信度: {best_confidence:.2f})")
            else:
                logger.warning(f"   ❌ 未找到匹配: {k_event.name}")
        
        return matched_events
    
    def _match_markets(
        self,
        matched_events: List[MatchedEvent]
    ) -> List[MatchedMarket]:
        """第二阶段: 市场匹配 (2:1)
        
        对于每个匹配的事件:
        - Kalshi 有 2 个市场（每个队伍一个）
        - Polymarket 有 1 个市场（包含两个队伍）
        - 创建 2 个 MatchedMarket，每个对应一个 Kalshi 市场
        """
        logger.info("-" * 40)
        logger.info("🎯 第二阶段: 市场匹配 (2:1)")
        logger.info("-" * 40)
        
        matched_markets = []
        
        for matched_event in matched_events:
            k_event = matched_event.kalshi_event
            p_event = matched_event.polymarket_event
            
            if not k_event or not p_event or not p_event.market:
                continue
            
            poly_market = p_event.market
            
            # 对于 Kalshi 的每个市场，创建一个 MatchedMarket
            for k_market in k_event.markets:
                team_name = k_market.team_name.upper()
                
                # 获取 Poly 市场对于该队伍的价格
                try:
                    poly_yes, poly_no = poly_market.get_price_for_team(team_name)
                except ValueError as e:
                    logger.warning(f"   ⚠️ {e}")
                    continue
                
                matched = MatchedMarket(
                    event_name=matched_event.event_name,
                    team_name=team_name,
                    kalshi_market=k_market,
                    polymarket_market=poly_market,
                    poly_yes_price=poly_yes,
                    poly_no_price=poly_no,
                    confidence=matched_event.confidence
                )
                matched_markets.append(matched)
                
                logger.info(f"   ✅ 市场匹配: {matched_event.event_name} - {team_name}")
                logger.debug(f"      Kalshi: Yes={k_market.yes_price:.2f}, No={k_market.no_price:.2f}")
                logger.debug(f"      Poly:   Yes={poly_yes:.2f}, No={poly_no:.2f}")
        
        return matched_markets
    
    def get_subscription_info(
        self,
        matched_markets: List[MatchedMarket]
    ) -> Tuple[List[str], List[str], Dict[str, List[MatchedMarket]]]:
        """获取 WebSocket 订阅信息
        
        Returns:
            kalshi_tickers: Kalshi 需要订阅的 ticker 列表
            polymarket_token_ids: Polymarket 需要订阅的 token ID 列表
            market_lookup: subscription_id -> [MatchedMarket] 的映射
        """
        kalshi_tickers = []
        polymarket_token_ids = []
        market_lookup: Dict[str, List[MatchedMarket]] = {}
        
        seen_kalshi = set()
        seen_poly = set()
        
        for mm in matched_markets:
            # Kalshi ticker
            k_id = mm.kalshi_market.market_id
            if k_id and k_id not in seen_kalshi:
                kalshi_tickers.append(k_id)
                seen_kalshi.add(k_id)
            
            # 添加到 lookup
            if k_id not in market_lookup:
                market_lookup[k_id] = []
            market_lookup[k_id].append(mm)
            
            # Polymarket token (根据队伍获取对应的 token)
            p_token = mm.polymarket_market.get_token_for_team(mm.team_name)
            if p_token and p_token not in seen_poly:
                polymarket_token_ids.append(p_token)
                seen_poly.add(p_token)
            
            # Poly token 也添加到 lookup
            if p_token:
                if p_token not in market_lookup:
                    market_lookup[p_token] = []
                market_lookup[p_token].append(mm)
        
        logger.info(f"📡 订阅信息: Kalshi {len(kalshi_tickers)} 个, Polymarket {len(polymarket_token_ids)} 个 token")
        
        return kalshi_tickers, polymarket_token_ids, market_lookup

File: rust-backend/src/core/matcher.rs
//! Event and Market Matcher
//!
//! Matching logic optimization:
//! - Event matching: Use canonical competitors + game date to match events on both platforms
//! - Market matching: 2:1 matching (two Kalshi markets correspond to one Poly market)
//!
//! Key points:
//! - Kalshi: One event has 2 markets (one for each team)
//! - Polymarket: One event has 1 market (contains prices for both teams)
//! - When matching, determine Poly price perspective based on team name

use std::collections::HashMap;
use tracing::{debug, info, warn};

use crate::models::{
    KalshiEvent, KalshiMarket, MatchedEvent, MatchedMarket, PolymarketEvent, PolymarketMarket,
};

/// Event matcher - two-stage matching
pub struct EventMatcher {
    /// Time tolerance in hours for matching
    #[allow(dead_code)]
    time_tolerance_hours: i64,
}

/// Subscription info returned by get_subscription_info
pub struct SubscriptionInfo {
    /// Kalshi tickers to subscribe
    pub kalshi_tickers: Vec<String>,
    /// Lookup map: subscription_id -> Vec<MatchedMarket index>
    pub market_lookup: HashMap<String, Vec<usize>>,
}

impl SubscriptionInfo {
    /// Create an empty subscription info
    pub fn empty() -> Self {
        Self {
            kalshi_tickers: Vec::new(),
            market_lookup: HashMap::new(),
        }
    }
}

impl EventMatcher {
    /// Create a new event matcher
    pub fn new(time_tolerance_hours: i64) -> Self {
        Self {
            time_tolerance_hours,
        }
    }

    /// Execute two-stage matching
    pub fn match_events_and_markets(
        &self,
        kalshi_events: &[KalshiEvent],
        _kalshi_markets: &[KalshiMarket],
        polymarket_events: &[PolymarketEvent],
        _polymarket_markets: &[PolymarketMarket],
    ) -> (Vec<MatchedEvent>, Vec<MatchedMarket>) {
        info!("============================================================");
        info!("🔍 Starting two-stage matching (non-split version)");
        info!("   Kalshi: {} events", kalshi_events.len());
        info!("   Polymarket: {} events", polymarket_events.len());
        info!("============================================================");

        // Stage 1: Event matching
        let matched_events = self.match_events(kalshi_events, polymarket_events);
        info!(
            "📊 Stage 1 complete: matched {} events",
            matched_events.len()
        );

        // Stage 2: Market matching (2:1)
        let matched_markets = self.match_markets(&matched_events);
        info!(
            "📊 Stage 2 complete: matched {} market pairs",
            matched_markets.len()
        );

        (matched_events, matched_markets)
    }

    /// Stage 1: Event matching
    fn match_events(
        &self,
        kalshi_events: &[KalshiEvent],
        polymarket_events: &[PolymarketEvent],
    ) -> Vec<MatchedEvent> {
        info!("----------------------------------------");
        info!("🎯 Stage 1: Event matching");
        info!("----------------------------------------");

        let mut matched_events = Vec::new();
        let mut used_poly_ids = std::collections::HashSet::new();

        // Build Polymarket event index: category + event_name -> [events].
        // Keeping sports separate prevents a coincidental competitor string from
        // ever matching an NBA event to a tennis event.
        let mut poly_index: HashMap<String, Vec<&PolymarketEvent>> = HashMap::new();
        for event in polymarket_events {
            let key = format!("{}:{}", event.category, event.name.to_uppercase());
            poly_index.entry(key).or_default().push(event);
        }

        // Find match for each Kalshi event
        for k_event in kalshi_events {
            let k_name = k_event.name.to_uppercase();
            let k_date = k_event.start_time.map(|t| t.date_naive());

            // Also check reversed name (e.g., MEM-LAL vs LAL-MEM)
            let k_name_reversed = {
                let parts: Vec<&str> = k_name.split('-').collect();
                if parts.len() == 2 {
                    Some(format!("{}-{}", parts[1], parts[0]))
                } else {
                    None
                }
            };

            let mut best_match: Option<&PolymarketEvent> = None;
            let mut best_confidence = 0.0_f64;

            // Look for exact match or reversed match
            for (name_to_check, is_reversed) in
                [(Some(&k_name), false), (k_name_reversed.as_ref(), true)]
            {
                let Some(name) = name_to_check else {
                    continue;
                };

                let index_key = format!("{}:{}", k_event.category, name);
                let candidates = poly_index
                    .get(&index_key)
                    .map(|v| v.as_slice())
                    .unwrap_or(&[]);

                for p_event in candidates {
                    if used_poly_ids.contains(&p_event.event_id) {
                        continue;
                    }

                    let p_date = p_event.start_time.map(|t| t.date_naive());

                    // Validate date
                    let confidence = match (k_date, p_date) {
                        (Some(kd), Some(pd)) => {
                            if kd != pd {
                                debug!(
                                    "   ❌ Date mismatch: {} ({}) vs {} ({})",
                                    k_event.name, kd, p_event.name, pd
                                );
                                continue;
                            }
                            if is_reversed {
                                0.95
                            } else {
                                1.0
                            }
                        }
                        _ => {
                            warn!(
                                "   ⚠️ Missing date: {} ({:?}) vs {} ({:?})",
                                k_event.name, k_date, p_event.name, p_date
                            );
                            if is_reversed {
                                0.65
                            } else {
                                0.7
                            }
                        }
                    };

                    if confidence > best_confidence {
                        best_confidence = confidence;
                        best_match = Some(p_event);
                    }
                }
            }

            if let Some(p_event) = best_match {
                if best_confidence >= 0.7 {
                    let matched = MatchedEvent {
                        event_name: k_event.name.clone(),
                        game_date: k_date, // Use Kalshi event's date
                        kalshi_event: Some(k_event.clone()),
                        polymarket_event: Some(p_event.clone()),
                        confidence: best_confidence,
                    };
                    matched_events.push(matched);
                    used_poly_ids.insert(p_event.event_id.clone());

                    info!(
                        "   ✅ Match: {} <-> {} (confidence: {:.2})",
                        k_event.name, p_event.name, best_confidence
                    );
                }
            } else {
                warn!("   ❌ No match found: {}", k_event.name);
            }
        }

        matched_events
    }

    /// Stage 2: Market matching (2:1)
    ///
    /// For each matched event:
    /// - Kalshi has 2 markets (one for each team)
    /// - Polymarket has 1 market (contains both teams)
    /// - Create 2 MatchedMarkets, each corresponding to one Kalshi market
    fn match_markets(&self, matched_events: &[MatchedEvent]) -> Vec<MatchedMarket> {
        info!("----------------------------------------");
        info!("🎯 Stage 2: Market matching (2:1)");
        info!("----------------------------------------");

        let mut matched_markets = Vec::new();

        for matched_event in matched_events {
            let Some(k_event) = &matched_event.kalshi_event else {
                continue;
            };
            let Some(p_event) = &matched_event.polymarket_event else {
                continue;
            };
            let Some(poly_market) = &p_event.market else {
                continue;
            };

            // For each Kalshi market, create a MatchedMarket
            for k_market in &k_event.markets {
                let team_name = k_market.team_name.to_uppercase();

                // Get Poly market prices for this team
                let (poly_yes, poly_no) = match poly_market.get_price_for_team(&team_name) {
                    Ok(prices) => prices,
                    Err(e) => {
                        warn!("   ⚠️ {}", e);
                        continue;
                    }
                };

                let matched = MatchedMarket {
                    event_name: matched_event.event_name.clone(),
                    team_name: team_name.clone(),
                    game_date: matched_event.game_date, // Inherit from MatchedEvent
                    kalshi_market: k_market.clone(),
                    polymarket_market: poly_market.clone(),
                    poly_yes_price: poly_yes,
                    poly_no_price: poly_no,
                    confidence: matched_event.confidence,
                };
                matched_markets.push(matched);

                info!(
                    "   ✅ Market match: {} - {}",
                    matched_event.event_name, team_name
                );
                debug!(
                    "      Kalshi: Yes={:.2}, No={:.2}",
                    k_market.yes_price, k_market.no_price
                );
                debug!("      Poly:   Yes={:.2}, No={:.2}", poly_yes, poly_no);
            }
        }

        matched_markets
    }

    /// Get WebSocket subscription info
    ///
    /// Polymarket US gateway markets are REST-polled, not CLOB-subscribed.
    /// The lookup therefore contains Kalshi tickers only.
    pub fn get_subscription_info(&self, matched_markets: &[MatchedMarket]) -> SubscriptionInfo {
        let mut kalshi_tickers = Vec::new();
        let mut market_lookup: HashMap<String, Vec<usize>> = HashMap::new();

        let mut seen_kalshi = std::collections::HashSet::new();

        for (idx, mm) in matched_markets.iter().enumerate() {
            // Kalshi ticker
            let k_id = &mm.kalshi_market.market_id;
            if !k_id.is_empty() && !seen_kalshi.contains(k_id) {
                kalshi_tickers.push(k_id.clone());
                seen_kalshi.insert(k_id.clone());
            }

            // Add to lookup
            market_lookup.entry(k_id.clone()).or_default().push(idx);

        }

        info!(
            "📡 Subscription information: Kalshi {} tickers; Polymarket US uses REST quotes",
            kalshi_tickers.len()
        );

        SubscriptionInfo {
            kalshi_tickers,
            market_lookup,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_test_poly_market() -> PolymarketMarket {
        PolymarketMarket {
            market_id: "poly-123".to_string(),
            market_slug: "lal-mem-2026-08-17".to_string(),
            event_name: "LAL-MEM".to_string(),
            team_a: "LAL".to_string(),
            team_b: "MEM".to_string(),
            price_a: 0.45,
            price_b: 0.55,
            team_a_position: crate::models::PolymarketPositionSide::Long,
            team_b_position: crate::models::PolymarketPositionSide::Short,
            start_time: None,
            volume: None,
        }
    }

    #[test]
    fn test_get_price_for_team() {
        let market = create_test_poly_market();

        // LAL is team_a
        let (yes, no) = market.get_price_for_team("LAL").unwrap();
        assert!((yes - 0.45).abs() < 0.001);
        assert!((no - 0.55).abs() < 0.001);

        // MEM is team_b
        let (yes, no) = market.get_price_for_team("MEM").unwrap();
        assert!((yes - 0.55).abs() < 0.001);
        assert!((no - 0.45).abs() < 0.001);

        // Case insensitive
        let (yes, _) = market.get_price_for_team("lal").unwrap();
        assert!((yes - 0.45).abs() < 0.001);

        // Invalid team
        assert!(market.get_price_for_team("BOS").is_err());
    }

    #[test]
    fn test_us_execution_for_competitor() {
        let market = create_test_poly_market();

        assert_eq!(
            market
                .us_execution_for_competitor("LAL")
                .unwrap()
                .position_side,
            crate::models::PolymarketPositionSide::Long
        );
        assert_eq!(market.us_execution_for_competitor("BOS"), None);
    }
}
